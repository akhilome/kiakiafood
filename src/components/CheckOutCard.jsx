import React from 'react';
import PropTypes from 'prop-types';

const CheckOutCard = ({
  id = 0, foodName = '', foodPrice = 0, buttonCallback,
}) => (
  <div data-foodid={id} className="food-card-checkout">
    <div className="food-card-checkout__details">
      <h4>{foodName}</h4>
      <p>
â‚¦
        {foodPrice}
      </p>
    </div>
    <button type="button" onClick={buttonCallback}>
      Remove
    </button>
  </div>
);

CheckOutCard.propTypes = {
  id: PropTypes.number.isRequired,
  foodName: PropTypes.string.isRequired,
  foodPrice: PropTypes.number.isRequired,
  buttonCallback: PropTypes.func.isRequired,
};

export default CheckOutCard;
