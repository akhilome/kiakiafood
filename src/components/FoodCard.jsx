import React from 'react';
import PropTypes from 'prop-types';

const FoodCard = ({
  foodImage, foodName, id, foodPrice, isAdmin,
}) => (
  <div className="food-card">
    <img className="food-image" src={foodImage} alt={foodName} />
    <div className="food-details">
      <div data-foodid={id} className="food-details__legend">
        <h4>{foodName}</h4>
        <p>
â‚¦
          {foodPrice.toLocaleString()}
        </p>
      </div>
      <div className="food-details__action">
        <button type="button">{isAdmin ? 'delete' : 'buy'}</button>
      </div>
    </div>
  </div>
);

FoodCard.propTypes = {
  foodImage: PropTypes.string.isRequired,
  foodName: PropTypes.string.isRequired,
  foodPrice: PropTypes.number.isRequired,
  isAdmin: PropTypes.bool.isRequired,
  id: PropTypes.number.isRequired,
};

export default FoodCard;
