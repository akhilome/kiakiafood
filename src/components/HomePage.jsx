import React from 'react';
import { Link } from 'react-router-dom';
import Nav from './Nav';

const HomePage = () => (
  <div>
    <Nav />
    <section className="hero">
      <h2>Quell Your Hunger in Minutes</h2>
      <Link to="/menu">
        <button type="button" className="btn-primary">
          Get Food ‚ü∂
        </button>
      </Link>
    </section>
  </div>
);

export default HomePage;
