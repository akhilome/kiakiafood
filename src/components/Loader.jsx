import React from 'react';

const Loader = () => (
  <div className="loader">
    <img src="https://i.imgur.com/S39IaLt.gif" alt="loading gif" />
  </div>
);

export default Loader;
