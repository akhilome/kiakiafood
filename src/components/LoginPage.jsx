import React from 'react';
import { Link } from 'react-router-dom';

const LoginPage = () => {
  return (
    <div>
      This is the login page! ğŸ”¥ See <Link to="/signup">signup</Link>
    </div>
  );
};

export default LoginPage;
