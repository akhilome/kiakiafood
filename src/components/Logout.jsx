import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { logout } from '../actions';

export class Logout extends Component {
  onLogoutClick = async (e) => {
    e.preventDefault();
    const { logout: logoutUser, history } = this.props;
    await logoutUser();
    return history.push('/login');
  };

  render() {
    return (
      <a onClick={this.onLogoutClick} href="!#">
        logout
      </a>
    );
  }
}

Logout.propTypes = {
  logout: PropTypes.func.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
};

export default connect(
  null,
  { logout },
)(withRouter(Logout));
