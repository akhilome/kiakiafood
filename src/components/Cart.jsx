import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import CheckOutCard from './CheckOutCard';
import { removeFromCart, checkout } from '../actions';

class Cart extends Component {
  onCheckoutClick = async () => {
    const { cart, checkout: placeOrder, history } = this.props;
    const foodItems = Object.keys(cart).map(Number);
    await placeOrder(foodItems);
    const {
      fetching: { fetching, errorMessage },
    } = this.props;
    if (!fetching && !errorMessage) return history.push('/order-history');
    return undefined;
  };

  onRemoveClick = (foodId) => {
    const { removeFromCart: removeItemFromCart } = this.props;
    removeItemFromCart(foodId);
  };

  render() {
    const { cart } = this.props;

    const cartItems = Object.keys(cart)
      .map(Number)
      .map(item => ({
        id: item,
        foodName: cart[item].foodName,
        foodPrice: cart[item].foodPrice,
      }));

    return (
      <div className="wrapper">
        <h2>Items To Be Purchased</h2>
        {cartItems.length ? (
          <Fragment>
            <section className="container checkout">
              {cartItems.map(item => (
                <CheckOutCard
                  key={item.id}
                  id={item.id}
                  foodName={item.foodName}
                  foodPrice={item.foodPrice}
                  buttonCallback={() => this.onRemoveClick(item.id)}
                />
              ))}
            </section>
            <div className="order-price">
              <p>
                <strong>total price</strong>
: â‚¦
                <span id="order-price">
                  {cartItems
                    .map(item => item.foodPrice)
                    .reduce((a, b) => a + b)
                    .toLocaleString()}
                </span>
              </p>
            </div>
            <button
              type="button"
              className="btn-primary"
              id="checkout"
              onClick={this.onCheckoutClick}
            >
              Checkout
            </button>
          </Fragment>
        ) : (
          <section className="container checkout">
            <div>No Items in cart</div>
          </section>
        )}
      </div>
    );
  }
}

Cart.propTypes = {
  cart: PropTypes.instanceOf(Object).isRequired,
  removeFromCart: PropTypes.func.isRequired,
  checkout: PropTypes.func.isRequired,
  history: PropTypes.oneOfType([PropTypes.object, PropTypes.func]).isRequired,
  fetching: PropTypes.instanceOf(Object).isRequired,
};

const mapStateToProps = state => ({
  cart: state.cart,
  fetching: state.fetching,
});

export default connect(
  mapStateToProps,
  { removeFromCart, checkout },
)(Cart);
