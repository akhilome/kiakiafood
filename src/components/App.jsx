import '../index.css';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import propTypes from 'prop-types';
import { checkAuthStatus } from '../actions';
import HomePage from './HomePage';
import LoginPage from './LoginPage';
import Signup from './SignupPage';
import NotFoundPage from './NotFoundPage';

export class App extends Component {
  componentDidMount() {
    const { checkAuthStatus: checkUserAuthStatus } = this.props;
    checkUserAuthStatus();
  }

  render() {
    return (
      <Router>
        <Switch>
          <Route path="/" exact component={HomePage} />
          <Route path="/login" component={LoginPage} />
          <Route path="/signup" component={Signup} />
          <Route component={NotFoundPage} />
        </Switch>
      </Router>
    );
  }
}

App.propTypes = {
  checkAuthStatus: propTypes.func.isRequired,
};

export default connect(
  null,
  { checkAuthStatus },
)(App);
