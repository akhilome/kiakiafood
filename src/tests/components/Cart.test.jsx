import React from 'react';
import { mount } from 'enzyme';

import { Cart } from '../../components/Cart';

describe('<Cart />', () => {
  const props = {
    cart: {
      1: {
        foodName: 'Sweet food',
        foodPrice: 829,
      },
      2: {
        foodName: 'Ugly plantain',
        foodPrice: 129,
      },
    },
    removeFromCart: jest.fn(),
    checkout: jest.fn(),
    history: { push: jest.fn() },
    fetching: { fetching: false, errorMessage: '' },
  };

  it('should render correctly', () => {
    const wrapper = mount(<Cart {...props} />);
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it('should remove item from cart successfully', () => {
    const wrapper = mount(<Cart {...props} />);
    const firstRemoveButton = wrapper.find('.food-card-checkout > button').first();
    firstRemoveButton.simulate('click');
    expect(props.removeFromCart).toHaveBeenCalledWith(1);
    wrapper.unmount();
  });

  it('should place order for food successfully', async () => {
    const wrapper = mount(<Cart {...props} />);
    const checkoutButton = wrapper.find('#checkout');
    await checkoutButton.simulate('click');
    expect(props.checkout).toHaveBeenCalledWith([1, 2]);
    expect(props.history.push).toHaveBeenCalledWith('/');
    wrapper.unmount();
  });

  it('should not redirect user if error occurs', async () => {
    const wrapper = mount(
      <Cart
        {...{
          ...props,
          fetching: { ...props.fetching, errorMessage: 'something went bonkers' },
        }}
      />,
    );
    const checkoutButton = wrapper.find('#checkout');
    await checkoutButton.simulate('click');
    expect(props.history.push).toHaveBeenCalledTimes(1);
    wrapper.unmount();
  });
});
