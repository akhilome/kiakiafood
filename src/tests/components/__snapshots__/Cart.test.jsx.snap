// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> should render correctly 1`] = `
<Cart
  cart={
    Object {
      "1": Object {
        "foodName": "Sweet food",
        "foodPrice": 829,
      },
      "2": Object {
        "foodName": "Ugly plantain",
        "foodPrice": 129,
      },
    }
  }
  checkout={[MockFunction]}
  fetching={
    Object {
      "errorMessage": "",
      "fetching": false,
    }
  }
  history={
    Object {
      "push": [MockFunction],
    }
  }
  removeFromCart={[MockFunction]}
>
  <div
    className="wrapper"
  >
    <h2>
      Items To Be Purchased
    </h2>
    <section
      className="container checkout"
    >
      <CheckOutCard
        buttonCallback={[Function]}
        foodName="Sweet food"
        foodPrice={829}
        id={1}
        key="1"
      >
        <div
          className="food-card-checkout"
          data-foodid={1}
        >
          <div
            className="food-card-checkout__details"
          >
            <h4>
              Sweet food
            </h4>
            <p>
              ₦
              829
            </p>
          </div>
          <button
            onClick={[Function]}
            type="button"
          >
            Remove
          </button>
        </div>
      </CheckOutCard>
      <CheckOutCard
        buttonCallback={[Function]}
        foodName="Ugly plantain"
        foodPrice={129}
        id={2}
        key="2"
      >
        <div
          className="food-card-checkout"
          data-foodid={2}
        >
          <div
            className="food-card-checkout__details"
          >
            <h4>
              Ugly plantain
            </h4>
            <p>
              ₦
              129
            </p>
          </div>
          <button
            onClick={[Function]}
            type="button"
          >
            Remove
          </button>
        </div>
      </CheckOutCard>
    </section>
    <div
      className="order-price"
    >
      <p>
        <strong>
          total price
        </strong>
        : ₦
        <span
          id="order-price"
        >
          958
        </span>
      </p>
    </div>
    <button
      className="btn-primary"
      id="checkout"
      onClick={[Function]}
      type="button"
    >
      Checkout
    </button>
  </div>
</Cart>
`;
